teos/gatekeeper.py

```diff
@@ -83,20 +89,46 @@ class Gatekeeper:
             database.
         registered_users (:obj:`dict`): A map of ``user_pk:user_info``.
         lock (:obj:`Lock`): A lock object to lock access to the Gatekeeper on updates.
-
     """
 
     def __init__(self, user_db, block_processor, subscription_slots, subscription_duration, expiry_delta):
         self.subscription_slots = subscription_slots
         self.subscription_duration = subscription_duration
         self.expiry_delta = expiry_delta
+        self.block_queue = Queue()
+        self.expired_users_cache = []
         self.block_processor = block_processor
         self.user_db = user_db
         self.registered_users = {
             user_id: UserInfo.from_dict(user_data) for user_id, user_data in user_db.load_all_users().items()
         }
         self.lock = Lock()
 
+        # Starts a child thread to take care of expiring subscriptions
+        Thread(target=self.manage_subscription_expiry, daemon=True).start()
+
+    def manage_subscription_expiry(self):
+        """
+        Manages the subscription expiry of the registered users. When a subscription reaches it end, it moves it to
+        the expiry cache and deleted it after ``GATEKEEPER_USERS_CACHE`` blocks, giving time to the ``Watcher`` and
+        the ``Responder`` to query the required data.
+        """
+        while True:
+            block_hash = self.block_queue.get()
+
+            # When the ChainMonitor is stopped, a final ChainMonitor.END_MESSAGE message is sent
+            if block_hash == ChainMonitor.END_MESSAGE:
+                break
+
+            with self.lock:
+                expired_users = self.get_expired_users(self.block_processor.get_block_count())
+                self.expired_users_cache.append(expired_users)
+                # Expired user deletion is delayed so the Watcher and Responder have enough time to query the associated
+                # data for the deletion of their data structures
+                if len(self.expired_users_cache) > USERS_CACHE_SIZE_BLOCKS:
+                    expired_users = self.expired_users_cache.pop(0)
+                    Cleaner.delete_expired_users(expired_users, self.registered_users, self.user_db)
```

Comment:
Is this still safe in case of a block re-org? I'd think it is, as long USERS_CACHE_SIZE_BLOCKS is significantly bigger than what we consider the re-org threshold (e.g. 10 >> 6). Agreed?

Reply:
It is not actually, since the users in the cache are removed from `registered_users` without checking if the subscription is still expired after `USERS_CACHE_SIZE_BLOCKS`. Since the cache is not being reworked in a case of a reorg, it could be the case that it enters on a weird state. 

If we need to check `register_users` at `expiry + expiry_delta` and also `USERS_CACHE_SIZE_BLOCKS` blocks later, it may be way simpler to just drop items from the `register_users` at `expiry + expiry_delta + USERS_CACHE_SIZE_BLOCKS` and get rid of the cache.

That would mean that, effectively, we would be giving the user an additional `USERS_CACHE_SIZE_BLOCKS` windows where it may be able to renew the subscription and keep the data undeleted. I think that may be worth the simplicity of the approach tho.

Reply:
More on this. Since we are already giving `expiry_delta` additional blocks to the user, we could rework the way that works. Subscription could expire exactly at `expiry`, and the repurpose `expire_delta` as the time we keep information around before deleting. That means that the user won't be able to query no send any data from the tower after `expiry`, but their data will be kept around `expiry_delta` more blocks, and then it will be permanently deleted.

That gets rid both of the cache and the additional constant.